File: Binary_Tree_Inorder_Traversal.cpp
/*94. Binary Tree Inorder Traversal Add to List
Description  Submission  Solutions
Total Accepted: 182020
Total Submissions: 408092
Difficulty: Medium
Contributors: Admin
Given a binary tree, return the inorder traversal of its nodes' values.

For example:
Given binary tree [1,null,2,3],
   1
    \
     2
    /
   3
return [1,3,2].

Note: Recursive solution is trivial, could you do it iteratively?

Hide Company Tags Microsoft
Hide Tags Tree Hash Table Stack
Hide Similar Problems (M) Validate Binary Search Tree (M) Binary Tree Preorder Traversal (H) Binary Tree Postorder Traversal (M) Binary Search Tree Iterator (M) Kth Smallest Element in a BST (H) Closest Binary Search Tree Value II (M) Inorder Successor in BST
*/

/*Use Stack to store the Tree node
always push the left node of the root into the stack, then add the left into the list/vector
then push the right node */
// C++
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    vector<int> inorderTraversal(TreeNode *root)
    {
        vector<int> ret;
        TreeNode *cur = root;
        stack<TreeNode *> S;

        while (cur || !S.empty())
        {
            while (cur)
            {
                S.push(cur);
                cur = cur->left;
            }
            cur = S.top();
            S.pop();
            ret.push_back(cur->val);
            cur = cur->right;
        }

        return ret;
    }
};

File: Binary_Tree_Inorder_Traversal.swift
/*
Given the root of a binary tree, return the inorder traversal of its nodes' values.



Example 1:


Input: root = [1,null,2,3]
Output: [1,3,2]
Example 2:

Input: root = []
Output: []
Example 3:

Input: root = [1]
Output: [1]
Example 4:


Input: root = [1,2]
Output: [2,1]
Example 5:


Input: root = [1,null,2]
Output: [1,2]


Constraints:

The number of nodes in the tree is in the range [0, 100].
-100 <= Node.val <= 100


Follow up:

Recursive solution is trivial, could you do it iteratively?
*/

/*
Solution 2:
Use Stack to store the Tree node
always push the left node of the root into the stack, then add the left into the list/vector
then push the right node

Time Complexity: O(n)
Space Complexity: O(logn)
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func inorderTraversal(_ root: TreeNode?) -> [Int] {
        guard let root = root else { return [] }
        var stack = [TreeNode?]()
        var inorder = [Int]()

        var cur: TreeNode? = root
        while cur != nil || !stack.isEmpty {
            while cur != nil {
                stack.append(cur)
                cur = cur?.left
            }
            cur = stack.removeLast()
            inorder.append(cur!.val)
            cur = cur?.right
        }
        return inorder
    }
}

/*
Solution 1:
recursively traverse node
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func inorderTraversal(_ root: TreeNode?) -> [Int] {
        var res = [Int]()
        _inorder(root, &res)
        return res
    }

    func _inorder(_ node: TreeNode?, _ res: inout [Int]) {
        guard let node = node else { return }
        _inorder(node.left, &res)
        res.append(node.val)
        _inorder(node.right, &res)
    }
}
